Fix build on OPENSSL_SYS_TANDEM and older POSIXes

t8m opened the pull request:
This should hopefully fix not just the build on NONSTOP platform but also older POSIX systems.

It also allows for passing `-DOPENSSL_NO_LOCALE` as a workaround to `./Configure` command.

This is alternative to #18233

Fixes #18232

t8m commented:
The master branch needs this fix as well, but it will require small tweaks. I'll create PR once this is reviewed.

rsbeckerca commented:
> @rsbeckerca Could you please test this one?

In progress. Could you include this define in the base `50-nonstop.conf` after this passes?

t8m commented:
> I'm getting the following issue, on every compile, passing -DOPENSSL_NO_LOCALE to configure:
> 
> ```
>   #  define OPENSSL_NO_LOCALE
>             ^
> "/home/ituglib/randall/openssl-3.0/./e_os.h", line 425: warning(140): macro redefined differently
> ```
> 
> passing in -DOPENSSL_NO_LOCALE= avoids this issue.

Fixed now. Also fixed the Windows build failure.

t8m commented:
> Could you include this define in the base `50-nonstop.conf` after this passes?

Why? I explicitly handle OPENSSL_SYS_TANDEM in e_os.h.

rsbeckerca commented:
> > Could you include this define in the base `50-nonstop.conf` after this passes?
> 
> Why? I explicitly handle OPENSSL_SYS_TANDEM in e_os.h.

A more general solution is not to hard-code TANDEM everywhere where there is an issue. I was fine putting it in `50-nonstop.conf` without any other ref. But ok. I'm retesting.

t8m commented:
At least for glibc 2.5 - which is the one on RHEL-5 it should work.

florinspatar commented:
> @florinspatar could you please try this patch?

This patch fixed the build on Solaris. For the other platforms, the build is still in progress.

rsbeckerca commented:
Tests pass on NonStop.

florinspatar commented:
> > @florinspatar could you please try this patch?
> 
> This patch fixed the build on Solaris. For the other platforms, the build is still in progress.

The build passed on HP-UX, AIX, Solaris and older Linux (RHEL5 & SLES10). I didn't tested on Windows at all.

levitte commented on providers/fips/fipsprov.c:
```diff
@@ -23,6 +23,7 @@
 #include "prov/seeding.h"
 #include "self_test.h"
 #include "internal/core.h"
+#include "e_os.h"
```
"Oh look what the cat dragged in..." 😄

beldmit replied:
Personally I don't like it (and probably it's worth considering a separate small header for locale-related defines in master) but making a duplication in 3.0 looks like a bad choice.

levitte replied:
> > "Oh look what the cat dragged in..." smile
> 
> Is that a good thing or a bad thing?

Some of us consider that header file an ugly duckling...  it drags in quite a lot of cruft.  But we don't have a sane alternative for now

paulidale replied:
:disappointed:

t8m replied:
> Some of us consider that header file an ugly duckling... it drags in quite a lot of cruft. But we don't have a sane alternative for now

IMO the solution for e_os.h would be to clean it up but not to remove it altogether.

rsbeckerca commented:
Is there an ETA for 3.0.4?

paulidale commented:
> Is there an ETA for 3.0.4?

Not at this stage.  However, given this break, soon.

paulidale commented:
The more of a mess this turns into, the less I like us dealing with locales.  I'm leaning towards reverting the original "fix" and instead assuming/requiring that the strings passed for algorithm names and properties be ASCII.

Question for OTC: **should we revert the original fix and rely on applications passing us sane strings?**

paulidale commented:
Also fixes #18244

beldmit commented:
The problem is not with ASCII but with a Turkish locale (and probably some other locales we are not aware of) that breaks the ASCII upper/lowercase assumptions.

t8m commented:
@paulidale this is not at all about the compared strings being ASCII or not. They are ASCII already but the Turkish locale makes it not work for any string containing i/I.

IMO there is no way to revert this apart from providing our own strcasecmp implementation that is pure ASCII.

t8m commented:
And given this PR fixes the problems discovered so far I would kindly ask you to remove the hold. If we later decide to do a different implementation of OPENSSL_strcasecmp() we can still do it without a complete revert.

And yeah we cannot fully revert the original fix as we cannot remove OPENSSL_strcasecmp() from the API anymore.

paulidale commented:
We've got macros that are the exact strings we want, is using them broken?  If not, it seems like a nice workaround that won't cause problems.

OSSL_DIGEST_NAME_RIPEMD160 might be broken, it has the dot less uppercase I and if spelt out as "ripemd160", it would be different in Turkish.

beldmit commented:
If we want users to be able to provide case-insensitive input, we are still broken.

rsbeckerca commented:
Retesting this fix at commit [9a6b5c5](https://github.com/openssl/openssl/pull/18241/commits/9a6b5c544333abb4101dd9b836112b64868f090f)

rsbeckerca commented:
LFTM

t8m commented:
Ping for second review so this can be merged once the hold is lifted. This fix needs to be done regardless of what we do with the locale initialization.

t8m commented:
Ping for second review so this can be merged once the hold is lifted. This fix needs to be done regardless of what we do with the locale initialization.

t8m commented:
Also @beldmit please re-approve.

beldmit commented:
My approval persists.

Tarnum-tst commented:
@t8m 
Proper location of e_os.h header file is 'include/internal/e_os.h' since d5f9166bacfb3757dfd6117310ad54ab749b11f9

rsbeckerca commented:
100% clean build/test on NonStop

t8m commented:
Closing as the commits were merged as part of #18293
